packages/vite/src/node/server/http.ts

```diff
@@ -159,3 +160,39 @@ async function getCertificate(config: ResolvedConfig) {
     return content
   }
 }
+
+export async function httpServerStart(
+  httpServer: HttpServer,
+  serverOptions: {
+    port: number
+    strict: boolean | undefined
+    host: string | undefined
+    logger: Logger
+  }
+): Promise<number> {
+  return new Promise((resolve, reject) => {
+    let { port, strict, host, logger } = serverOptions
+
+    const onError = (e: Error & { code?: string }) => {
+      if (e.code === 'EADDRINUSE') {
+        if (strict) {
+          httpServer.removeListener('error', onError)
+          reject(new Error(`Port ${port} is already in use`))
+        } else {
+          logger.info(`Port ${port} is in use, trying another one...`)
+          httpServer.listen(++port, host)
+        }
+      } else {
+        httpServer.removeListener('error', onError)
+        reject(e)
```

Comment:
I think this is okay (using `new Promise`) cause the `reject` is called in the inner `onError` function.